lib/hotel.rb

```diff
@@ -0,0 +1,139 @@
+      @all_reservations = []
+      @block_reservations = []
+    end
+
+    # I can reserve a room for a given date range
+    def create_reservation(name, check_in, check_out)
+      valid?(name) # Checks name input
+      available_rooms = check_avail_rooms_for(check_in, check_out) # Returns array of all available rooms and checks valid dates input
+      if rooms_available?(available_rooms) # Your code should raise an exception when asked to reserve a room that is not available
+        assigned_room = available_rooms[0] # Assign the first available room
+        reservation = Reservation.new(name, assigned_room, check_in, check_out)
+        @all_reservations << reservation
+      end
+    end
+
+    def find_reservation(name)
+      valid?(name)
+      @all_reservations.each do |reservation|
+        if reservation.class == BookingSystem::Reservation
+          if reservation.name == name
+            return reservation
+          end
+        end
+      end
+      raise ArgumentError.new("No reservation was found")
+    end
+
+    # I can create a block of rooms
+    def reserve_block(reserved_for, check_in, check_out, num_of_rooms)
+      valid?(reserved_for)
+      check_room_input(num_of_rooms) # Raised to UI. Max of 5 rooms and no less than 1
+      available_rooms = check_avail_rooms_for(check_in, check_out)
+      if rooms_available?(available_rooms) # Will raise an ArgumentError for UI
+        reserved_room_nums = available_rooms[0..num_of_rooms - 1] # Takes the first 5 rooms from all available rooms
+        new_block = Block.new(reserved_for, check_in, check_out, reserved_room_nums)
+        # Return only one block reservation and pass it into the @all_reservations array so that they cannot access the rooms
+        @all_reservations << new_block
+        @block_reservations << new_block
+        return new_block # Returns all reserved rooms numbers for a block
+      end
+    end
+
+    def find_block(reserved_name)
+      valid?(reserved_name)
+      @block_reservations.each do |block|
+        if block.reserved_for == reserved_name
+          return block
+        end
+      end
+      raise ArgumentError.new("No block was found")
+    end # def
+
+    # As an administrator, I can reserve a room from within a block of rooms
+    def reserve_room_in_block(reserved_name, num_to_book)
+      valid?(reserved_name)
+      found_block = find_block(reserved_name)
+      check_num_input(found_block, num_to_book) # Raise ArgumentError to UI
+      avail_rooms = avail_rooms_in_block(reserved_name) # Check which room numbers are available in the block
+      now_reserved_in_block = avail_rooms[0..num_to_book - 1] # Book number of requested rooms in the current block
+      remaining_rooms = avail_rooms - now_reserved_in_block # Remaining rooms in requested block
+      found_block.update_rooms_in_block(remaining_rooms, now_reserved_in_block) # Updates available and reserved rooms in the requested block
```

Comment:
This is an example of tight coupling between `Hotel` and `Block`. `Hotel` finds the block, pulls out its rooms, selects some of the rooms and then updates the `Block`.

To decouple them, you might write a method `Block#reserve_rooms(num_to_book)` that does all the work of manipulating the list of rooms, and call that method here. That way `Hotel` does't need to rely on how rooms are stored in the `Block`.